pytorch_lightning/profiler/pytorch.py

```diff
@@ -102,202 +140,207 @@ def __init__(
-            self.on_train_start(local_rank=local_rank)
-            self.on_train_start = super().on_train_start
-
-    def on_train_start(self, local_rank: Optional[str] = None):
-        self.local_rank = local_rank
+        self.recording_map: Dict[str, record_function] = {}
+        self.profiler: Optional[_PROFILER] = None
+        self.function_events: Optional[EventList] = None
+        self._profiler_instantiated: bool = False
 
-        # when logging to `log.info`, only perform profiling on rank 0
-        if local_rank != 0 and self.output_fname is None:
-            self.wrap_functions_into_rank_zero_only()
+        super().__init__(output_filename=output_filename, local_rank=local_rank)
 
-        if self.output_fname:
-            if local_rank is not None:
-                if '.txt' not in self.output_fname:
-                    raise MisconfigurationException("Log file should be .txt file.")
+    def __deprecation_check(self, profiled_functions: List[str] = [], record_functions: List[str] = []) -> List[str]:
+        if profiled_functions is not None:
+            rank_zero_warn(
+                "`PyTorchProfiler.profiled_functions` has been renamed to"
+                " `record_functions` in v1.3 and will be removed in v1.5", DeprecationWarning
+            )
+            if (len(record_functions) == 0 or len(profiled_functions) == 0):
+                record_functions += profiled_functions
+            else:
+                raise MisconfigurationException(
+                    "You set `PytorchProfiler.profiled_functions` and `PyTorchProfiler.record_functions`."
+                    "  Please use only the later."
+                )
+        return record_functions
 
-                self.output_fname = self.output_fname.replace(".txt", f"_{self.local_rank}.txt")
+    def on_train_start(self, local_rank: Optional[int] = None, log_dir: Optional[str] = None) -> None:
+        super().on_train_start(local_rank=local_rank, log_dir=log_dir)
 
-            fs = get_filesystem(self.output_fname)
-            self.output_file = fs.open(self.output_fname, "w")
+        # if the user didn't provide `path_to_export_trace`,
+        # set it as TensorBoardLogger log_dir if exists
+        if self.path_to_export_trace is None:
+            self.path_to_export_trace = log_dir
 
-        streaming_out = [self.output_file.write] if self.output_file else [log.info]
-        super().__init__(output_streams=streaming_out)
+        # when logging to `log.info`, only perform profiling on rank 0
+        if local_rank is not None and local_rank > 0 and self.output_fname is None:
+            self._rank_zero_only_wrap()
 
-    def wrap_functions_into_rank_zero_only(self):
+    def _rank_zero_only_wrap(self) -> None:
         self.start = rank_zero_only(self.start)
         self.stop = rank_zero_only(self.stop)
         self.summary = rank_zero_only(self.summary)
         self.describe = rank_zero_only(self.describe)
```

Comment:
why profile only rank 0? i think it'd be helpful to have profiling run on all ranks, as oftentimes we'd need to find the max time spent across all the ranks for each step to figure out where the bottleneck is

Reply:
Notice how this wrapping is only done if `self.output_fname is None`

So we profile for all ranks if using a file, otherwise (stdout) just rank 0